### PR TITLE
Config validators

### PR DESCRIPTION
**Pull request by Vladislav-CS**
### Description
This PR offers attributes for validating config values.
It supports making own validators (just make a new attribute that implements IValidator)
```cs
[LessThan(10)]
[GreaterThan(1)]
[NonNegative]
[NonPositive]
[PossibleValues<RoleTypeId>(RoleTypeId.ClassD, RoleTypeId.Scientist)]
public SomeType SomeValue { get; set; }
```
### Changes
* Update C# version to 11 (attribute with generic type)
* New interface IValidator with validate method
* Default validators: GreaterThan, LessThan, NonNegative, NotPositive, PossibleValues
* Log how many validations are passed
### Related suggestions
* https://discord.com/channels/656673194693885975/663978314683252777/1149086332405170299
### Tasks
- [x] Make sure it works

### CODE REVIEW

File: EXILED.props [hunk 1]
```diff
@@ -7,7 +7,7 @@
 
   <PropertyGroup Condition="$(BuildProperties) == '' OR $(BuildProperties) == 'true'">
     <TargetFramework>net48</TargetFramework>
-    <LangVersion>9.0</LangVersion>
+    <LangVersion>11.0</LangVersion>
```
Comment (louis1706) [inline EXILED.props:10]: No

**Review by louis1706 — CHANGES_REQUESTED** [review]

File: EXILED.props (hunk 1 above)
Comment (Vladislav-CS) [reply EXILED.props:10]: Why?

Comment (louis1706) [reply EXILED.props:10]: > Why?

We need more discussion about updating C# i will not allow if not everyone is okay with it

**Review by Ruemena — CHANGES_REQUESTED** [review]
should make everything a float instead of an int (cast to float), and need greaterthaneq and lessthaneq

File: Exiled.API/Features/Attributes/Validators/PossibleValuesAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+            Values = values;
+        }
+
+        /// <summary>
+        /// Gets the values.
+        /// </summary>
+        public T[] Values { get; }
+
+        /// <inheritdoc/>
+        public bool Validate(object value) => value is T type && Values.Contains(type);
```
Comment (louis1706) [inline Exiled.API/Features/Attributes/Validators/PossibleValuesAttribute.cs:36]: if Values is empty and T is Enum verify than the Enum.IsDefined(T.Type, value) // something like that

File: Exiled.Loader/ConfigManager.cs [hunk 3]
```diff
@@ -137,7 +136,39 @@ public static IConfig LoadSeparatedConfig(this IPlugin<IConfig> plugin)
...
+                    else
+                    {
+                        validations++;
+                    }
+                }
+            }
+
+            if (validations > 0)
+            {
+                Log.Info($"Successfully passed {validations} validations");
```
Comment (louis1706) [inline Exiled.Loader/ConfigManager.cs:171]: change this to a Debug

**Review by louis1706 — CHANGES_REQUESTED** [review]

File: Exiled.Loader/ConfigManager.cs (hunk 3 above)
Comment (ALEXWARELLC) [reply Exiled.Loader/ConfigManager.cs:171]: I'll get that done once I get a chance.

**Comment by ALEXWARELLC** [comment]
@louis1706 What other changes?

**Comment by ALEXWARELLC** [comment]
What’s still needed to do?

**Comment by louis1706** [comment]
[PossibleValues<RoleTypeId>()]

Does this authorize all value from RoleTypeId ?

File: Exiled.Loader/ConfigManager.cs [hunk 4]
```diff
@@ -16,8 +16,8 @@ namespace Exiled.Loader
     using API.Enums;
     using API.Extensions;
     using API.Interfaces;
-
     using Exiled.API.Features;
+    using Exiled.API.Features.Pools;
```
Comment (ALEXWARELLC) [inline Exiled.Loader/ConfigManager.cs:20]: I believe this is not needed as it does not exist. Hence the build compilation error in the CI.

**Review by ALEXWARELLC — COMMENTED** [review]
CI issue fix (Maybe)

I'm gonna create a PR in the source fork to fix this if it's ok.

File: Exiled.Loader/ConfigManager.cs (hunk 4 above)
Comment (ALEXWARELLC) [reply Exiled.Loader/ConfigManager.cs:20]: https://github.com/Vladislav-CS/EXILED/pull/3

**Review by ALEXWARELLC — COMMENTED** [review]
Sadge.

**Comment by ALEXWARELLC** [comment]
> ```cs
> [PossibleValues<RoleTypeId>()]
> ```
> 
> Does this authorize all value from RoleTypeId ?

I can't test it since I am unable to build a Exiled.Loader.dll bc of Visual Studio being dumb. But I will keep trying.

File: Exiled.API/Features/Attributes/Validators/GreaterThanAttribute.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Property)]
+    public sealed class GreaterThanAttribute : Attribute, IValidator
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="GreaterThanAttribute"/> class.
+        /// </summary>
+        /// <param name="number">A number the value should be greater.</param>
+        /// <param name="isIncluded">Whether or not <paramref name="number"></paramref> is included.</param>
+        public GreaterThanAttribute(int number, bool isIncluded = false)
```
Comment (ALEXWARELLC) [inline Exiled.API/Features/Attributes/Validators/GreaterThanAttribute.cs:25]: float?

File: Exiled.API/Features/Attributes/Validators/LessThanAttribute.cs [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Property)]
+    public sealed class LessThanAttribute : Attribute, IValidator
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="LessThanAttribute"/> class.
+        /// </summary>
+        /// <param name="number">A number the value should be less.</param>
+        /// <param name="isIncluded">Whether or not <paramref name="number"></paramref> is included.</param>
+        public LessThanAttribute(int number, bool isIncluded = false)
```
Comment (ALEXWARELLC) [inline Exiled.API/Features/Attributes/Validators/LessThanAttribute.cs:25]: float?

File: Exiled.API/Features/Attributes/Validators/GreaterThanAttribute.cs [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        {
+            Number = number;
+            IsIncluded = isIncluded;
+        }
+
+        /// <summary>
+        /// Gets the number.
+        /// </summary>
+        public int Number { get; }
```
Comment (ALEXWARELLC) [inline Exiled.API/Features/Attributes/Validators/GreaterThanAttribute.cs:34]: float?

File: Exiled.API/Features/Attributes/Validators/LessThanAttribute.cs [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+        {
+            Number = number;
+            IsIncluded = isIncluded;
+        }
+
+        /// <summary>
+        /// Gets the number.
+        /// </summary>
+        public int Number { get; }
```
Comment (ALEXWARELLC) [inline Exiled.API/Features/Attributes/Validators/LessThanAttribute.cs:34]: float?

**Review by ALEXWARELLC — CHANGES_REQUESTED** [review]
Could these be switched to floats? This would give (maybe) int support natively.

File: Exiled.API/Features/Attributes/Validators/GreaterThanAttribute.cs (hunk 5 above)
Comment (ALEXWARELLC) [reply Exiled.API/Features/Attributes/Validators/GreaterThanAttribute.cs:25]: Resolved

File: Exiled.API/Features/Attributes/Validators/LessThanAttribute.cs (hunk 6 above)
Comment (ALEXWARELLC) [reply Exiled.API/Features/Attributes/Validators/LessThanAttribute.cs:25]: Resolved

File: Exiled.API/Features/Attributes/Validators/LessThanAttribute.cs (hunk 8 above)
Comment (ALEXWARELLC) [reply Exiled.API/Features/Attributes/Validators/LessThanAttribute.cs:34]: Resolved

File: Exiled.API/Features/Attributes/Validators/GreaterThanAttribute.cs (hunk 7 above)
Comment (ALEXWARELLC) [reply Exiled.API/Features/Attributes/Validators/GreaterThanAttribute.cs:34]: Resolved

**Comment by ALEXWARELLC** [comment]
@NaoUnderscore Is this PR 'Do Not Merge'?

**Comment by ALEXWARELLC** [comment]
After testing, here are the issues I have noted:
- YamlDotNet 13.7.1 is required.
- Validator throws 2 errors instead of the intentional 1 error.

More notes may be added as we discover more issues.

**Review by louis1706 — APPROVED** [review]

**Comment by moddedmcplayer** [comment]
The docs are lacking.

**Comment by ALEXWARELLC** [comment]
> The docs are lacking.

We'll work on some documentation asap.

**Comment by moddedmcplayer** [comment]
> ALEXWARELLC

[Check your fork's PRs.](https://github.com/Vladislav-CS/EXILED/pull/5)

**Comment by Vladislav-CS** [comment]
> > ALEXWARELLC
> 
> [Check your fork's PRs.](https://github.com/Vladislav-CS/EXILED/pull/5)

Take a look at my review

**Review by ALEXWARELLC — APPROVED** [review]
👍

**Pull request merged**